app/components/scan-qr/media-stream-provider.tsx

```diff
@@ -0,0 +1,330 @@
+        video.setAttribute("playsinline", "playsinline");
+        try {
+          video.srcObject = mediaStream;
+        } catch (e) {
+          sendNotification({
+            title: "camera acces error",
+            message: "smothing went wrong",
+            icon: { name: "trash", variant: "error" },
+          });
+        }
+        await video.play().catch((error) => {
+          sendNotification({
+            title: "camera acces error",
+            message: error.toString(),
+            icon: { name: "trash", variant: "error" },
+          });
+        });
+        playVideoState();
+
+        setSupportedConstraints(
+          navigator.mediaDevices?.getSupportedConstraints()
+        );
+
+        const newMediaDevices =
+          await navigator.mediaDevices?.enumerateDevices();
+        setMediaDevices(newMediaDevices);
+        setVideoDevices(
+          newMediaDevices.filter((device) => device.kind === "videoinput")
+        );
+      } catch (error: any) {
+        sendNotification({
+          title: "camera acces error",
+          message: error.toString(),
+          icon: { name: "trash", variant: "error" },
+        });
+      }
+    },
+    [playVideoState, setCurrentVideoDeviceId]
+  );
+
+  const stopMediaStream = useCallback<
+    MediaStreamContextValue["stopMediaStream"]
+  >(() => {
+    try {
+      const video = videoRef.current;
+      if (!video) return;
+
+      const mediaStream = video.srcObject as MediaStream;
+      mediaStream?.getTracks().forEach((track) => {
+        track.stop();
+      });
+      video.srcObject = null;
+
+      stopVideoState();
+
+      setVideoTrackInfo(undefined);
+      setCurrentVideoDeviceId("");
+
+      alert("Stop Media Stream");
+    } catch (error: any) {
+      sendNotification({
+        title: "camera acces error",
+        message: error.toString(),
+        icon: { name: "trash", variant: "error" },
+      });
+      alert(error.toString());
+    }
+  }, [videoRef, stopVideoState]);
+
+  const onChangeConstraints = useCallback<
+    MediaStreamContextValue["onChangeConstraints"]
+  >(
+    async (constraints) => {
+      const video = videoRef.current;
+      if (!video) return;
+
+      const mediaStream = video.srcObject as MediaStream;
+      const videoTrack = mediaStream.getVideoTracks()[0];
+      if (!videoTrack) return;
+
+      const advanced: MediaTrackConstraintSet[] = [];
+
+      const deviceId = videoTrack.getSettings().deviceId;
+
+      advanced.push({
+        deviceId: constraints?.deviceId || deviceId,
+      });
+
+      if (constraints?.height) {
+        advanced.push({
+          height: constraints.height,
+        });
+      }
+
+      if (constraints?.width) {
+        advanced.push({
+          width: constraints.width,
+        });
+      }
+
+      if (constraints?.aspectRatio) {
+        advanced.push({
+          aspectRatio: constraints.aspectRatio,
+        });
+      }
+
+      alert(JSON.stringify(constraints));
```

Comment:
This shouldnt be here right? We have the sendNotification thing already.

Reply:
added new commit